File: Computer Networks/C3310-Week7-program-timeouts/tcp_client.py
#!/usr/bin/env python

"""
TCP echo client program for ANU COMP3310.

Run with
    python tcp_client.py [address [port]]

Written by Hugh Fisher (u9100925) and Felix Friedlander (u6675843), ANU,
2024.
CC0: This code is dedicated to the public domain.
This code may be freely copied and modified for any purpose.
https://creativecommons.org/publicdomain/zero/1.0/
"""

from socket import socket, AF_INET, SOCK_STREAM
from sys import argv

service_host = "127.0.0.1"
service_port = 3310

def loop(host, port):
    """
    Read lines of input, send them to the echo server, and wait for and print
    the server's response.
    """

    sock = socket(AF_INET, SOCK_STREAM)
    sock.connect((host, port))

    connected_host, connected_port = sock.getpeername()
    print(f"Connected to {connected_host}:{connected_port}")

    while True:
        try:
            line = input()
        except EOFError:
            break

        send_request(sock, line + "\n")
        read_reply(sock)

    print("Closing socket")
    sock.close()

def send_request(sock, request):
    """Send our request to server"""

    with sock.makefile("w") as sock_file:
        sock_file.writelines([request])
        sock_file.flush()
        print("Sent request to server...")


def read_reply(sock):
    """Read and print a response from the server"""

    with sock.makefile("r") as sock_file:
        reply = sock_file.readline()
        print("Received response:", reply, end="")

if __name__ == "__main__":
    host = argv[1] if len(argv) >= 2 else service_host
    port = argv[2] if len(argv) >= 3 else service_port

    loop(host, port)
    print("Done.")

File: Computer Networks/C3310-Week7-program-timeouts/TCPClient.java
/**
 * TCP echo client program for ANU COMP3310.
 *
 * Run with
 *     java TCPClient [address [port]]
 *
 * CC0: This code is dedicated to the public domain.
 * This code may be freely copied and modified for any purpose.
 * https://creativecommons.org/publicdomain/zero/1.0/
 *
 * @author Hugh Fisher (u9011925), ANU, 2024.
 * @author Felix Friedlander (u6675843), ANU, 2024.
 */

import java.io.*;
import java.net.*;

public class TCPClient {

	public static void main(String[] args) throws IOException {
		String serviceHost = "127.0.0.1";
		int servicePort = 3310;

		/*
		 * For arguments any more complicated than this,
		 * better to use a library.
		 */
		if (args.length > 0) {
			serviceHost = args[0];
			if (args.length > 1)
				servicePort = Integer.parseInt(args[1]);
		}

		loop(serviceHost, servicePort);
		System.out.println("Done.");
	}

	/**
	 * Read lines of input, send them to the echo server, and wait for
	 * and print the server's response.
	 */
	protected static void loop(String host, int port) throws IOException {
		Socket sock = new Socket(host, port);
		InetSocketAddress remote =
			(InetSocketAddress)sock.getRemoteSocketAddress();
		System.out.printf("Connected to %s:%d\n",
			remote.getAddress().getHostAddress(), remote.getPort());

		BufferedReader in =
			new BufferedReader(new InputStreamReader(sock.getInputStream()));
		PrintWriter out = new PrintWriter(sock.getOutputStream(), true);

		BufferedReader stdin =
			new BufferedReader(new InputStreamReader(System.in));

		while (true) {
			String line = stdin.readLine();
			if (line == null)
				break;

			sendRequest(out, line);
			readReply(in);
		}
		System.out.println("Closing socket");
		sock.close();
	}

	/**
	 * Send a request to the server.
	 */
	protected static void sendRequest(PrintWriter out, String request)
		throws IOException {

		out.println(request);
		System.out.println("Sent request to server...");
	}

	/**
	 * Read and print a response from the server.
	 */
	protected static void readReply(BufferedReader in) throws IOException {
		String reply = in.readLine();
		System.out.print("Received response: ");
		System.out.println(reply);
	}

}
